lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -2547,9 +2778,65 @@ usePrespecialized(SILOptFunctionBuilder &funcBuilder, ApplySite apply,
     auto fn = lookupOrCreatePrespecialization(funcBuilder, refF, name, reInfo);
     if (!specializationAvail.isAlwaysAvailable())
       fn->setAvailabilityForLinkage(specializationAvail);
-    return fn;
+
+    result.setFunction(fn);
+
+    auto fnType = fn->getLoweredFunctionType();
+    if (isAnyObjectSpecialization && fnType != apply.getSubstCalleeType()) {
+      auto &M = fn->getModule();
+      auto expansion = fn->getTypeExpansionContext();
+      auto calleeTy = apply.getSubstCalleeType();
+      auto substConv = apply.getSubstCalleeConv();
+      auto resultType =
+          fn->getConventions().getSILResultType(fn->getTypeExpansionContext());
+
+      for (auto pair : llvm::enumerate(apply.getArgumentOperands())) {
+        auto param = pair.value().get();
+        if (pair.index() < substConv.getSILArgIndexOfFirstParam() &&
+            resultType != param->getType()) {
+          assert(!result.hasIndirectResultType());
+          if (apply.getKind() == ApplySiteKind::TryApplyInst) {
+            result.setIndirectResultType(resultType.getObjectType());
+          } else {
+            result.setIndirectResultType(param->getType().getObjectType());
+          }
+          continue;
+        }
+
+        unsigned paramIdx =
+            pair.index() - substConv.getSILArgIndexOfFirstParam();
+
+        auto newParamType = fnType->getParameters()[paramIdx].getArgumentType(
+            M, fnType, expansion);
+        auto oldParamType = calleeTy->getParameters()[paramIdx].getArgumentType(
+            M, fnType, expansion);
+        if (newParamType != oldParamType) {
+          result.addParameterTypeReplacement(paramIdx, newParamType);
+        }
+      }
+
+      auto newConv = fn->getConventions();
+      for (auto pair : llvm::enumerate(substConv.getYields())) {
+        auto index = pair.index();
+        auto newType =
+            newConv.getYields()[index].getYieldValueType(M, fnType, expansion);
+        auto oldType = pair.value().getYieldValueType(
+            M, calleeTy, apply.getCalleeFunction()->getTypeExpansionContext());
+
+        if (oldType != newType) {
+          result.addYieldTypeReplacement(index, oldType);
+        }
+      }
+
+      if (!result.hasIndirectResultType() &&
+          resultType != apply.getType().getObjectType()) {
+        result.setResultType(apply.getType().getObjectType());
+      }
+    }
```

Comment:
I think a good test case to exercise whether this code works with multiple indirect result types would be something like:
```
@_specialize(exported: true, where T @ _Class, V @ _Class)
public func t<T, V>(_ t: T, _ v: V, _ x: Int64)-> (V, Int64, T) {
    return (v, x, t)
}

public class MyClass {}
public class YourClass {}
public func useT(_ c: MyClass, _ d: YourClass, _ x: Int64) {
 print(t(c, d, x))
}
```

Reply:
Would it make sense to move this code into a method of `SpecializedFunction` ?

Reply:
Added a test case for the case you are mentioning above and fixed the code so it passes now.